specs/language/resources.tex

```diff
@@ -663,19 +664,241 @@
+\end{grammar}
+
+\p The name of the \texttt{cbuffer} declaration group cannot be referenced from within the translation unit and is not required to be unique.
+
+\p Variable declarations in the \texttt{cbuffer} declaration group are called \textit{shader constants}.
+
+\p Shader constants can be referenced from anywhere in the translation unit after they are declared by directly using the declaration name. This implies that all shader constants declared in a translation unit must have unique names, even though they might be declared in different \texttt{cbuffer} declaration groups.
+
+\p Variable declarations in the \texttt{cbuffer} declaration group cannot have \texttt{groupshared} or \texttt{static} variable modifiers.
+
+\p Any declaration in the \texttt{cbuffer} declaration group that is not a
+\textit{namespace-declaration} or a \textit{variable-declaration} is going to be
+processed as if it was declared outside of the group at the
+global scope.
+
+\p Nesting of \texttt{cbuffer} declaration groups is not allowed.
+
+\p Any \textit{namespace-declaration} in the \texttt{cbuffer} declaration group can contain only declarations allowed by \textit{cbuffer-declaration}.
+
+\p For example:
+
+\begin{HLSL}
+  cbuffer MyConstants {
+    float4 CameraPos;
+  };
+
+  float4 getCameraPosition() {
+    return CameraPos;
+  }
+\end{HLSL}
+
+\Sub{Constant Buffer Class}{Resources.cnbuf.cbclass}
+
+\p Another way of declaring constant buffers is by using the
+\texttt{ConstantBuffer<T>} resource class.
+
+\p The template parameter \texttt{T} must be a class type (\ref{Classes}). 
+
+\begin{HLSL}
+  template <typename T>
+   class ConstantBuffer {
+   public:
+     ConstantBuffer();
+     ConstantBuffer(const ConstantBuffer &buf);
+     ConstantBuffer &operator=(ConstantBuffer &buf);
+  };
+\end{HLSL}
+
+\p The \textit{shader constants} are the fields of the class type \texttt{T}.
+They can be referenced from anywhere within the translation unit after the
+\texttt{ConstantBuffer<T>} declaration as if they were fields of the
+\texttt{ConstantBuffer<T>} class.
+
+\p For example:
+
+\begin{HLSL}
+  struct MyConstants {
+    float4 CameraPos;
+  };
+
+  ConstantBuffer<MyConstants> CB;
+
+  float4 getCameraPosition() {
+    return CB.CameraPos;
+  }
+\end{HLSL}
+
+\p The layout rules for constant buffer declared with the
+\texttt{ConstantBuffer<T>} syntax are the same as for named \texttt{cbuffer}
+declaration groups.
+
+\p \texttt{ConstantBuffer<T>} can be defined at local scope to represent a local
+reference to a constant buffer that can be associated with a global buffer when
+assigned. The reference must be resolvable to a unique global buffer declaration
+before use.
+
+\p Constant buffers cannot be implicitly nor explicitly cast from one type to
+another. This means that local \texttt{ConstantBuffer} can only be assigned from
+\texttt{ConstantBuffer} with the same template type \texttt{T}.
+
+\Sub{Constant Buffer Layout}{Resources.cnbuf.lay}
```

Comment:
We probably shouldn't describe the layout as part of the language. This is really a constraint of DXIL that SPIR-V doesn't have, and we're almost certainly going to want to change this in a future shader model.

The language you've written here is great though. We should take this language and capture it in the DXIL specification.

Reply:
What about `packoffset` annotations? They are part of the language and define specific offset in the constant
buffer.